libctru/source/synchronization.c

```diff
@@ -82,6 +90,72 @@ void LightLock_Lock(LightLock* lock)
 	__dmb();
 }
 
+int LightLock_LockTimeout(LightLock* lock, s64 timeout_ns)
+{
+	s32 val;
+	bool bAlreadyLocked;
+
+	// Try to lock, or if that's not possible, increment the number of waiting threads
+	do
+	{
+		// Read the current lock state
+		val = __ldrex(lock);
+		if (val == 0) val = 1; // 0 is an invalid state - treat it as 1 (unlocked)
+		bAlreadyLocked = val < 0;
+
+		// Calculate the desired next state of the lock
+		if (!bAlreadyLocked)
+			val = -val; // transition into locked state
+		else
+			--val; // increment the number of waiting threads (which has the sign reversed during locked state)
+	} while (__strex(lock, val));
+
+	// If the lock is held by a different thread, we need to do some timing initialization
+	if (bAlreadyLocked)
+	{
+		struct timespec startTime;
+		clock_gettime(CLOCK_MONOTONIC, &startTime);
```

Comment:
Prefer to use native HOS calls instead of libc from within system support code (applies to the entire file)